applications/vanilla/controllers/api/DiscussionsApiController.php

```diff
@@ -645,4 +667,68 @@ public function put_bookmark($id, array $body) {
         $result = $this->discussionByID($id);
         return $out->validate($result);
     }
+
+    /**
+     * Set canonical url for a discussion.
+     *
+     * @param int $id The ID of the discussion.
+     * @param array $body The request body.
+     *       Ex: ["canonicalUrl" => "https://mydomain.com/some+path/"]
+     * @throws NotFoundException If unable to find the discussion.
+     * @return array
+     */
+    public function put_canonical($id, array $body) {
+        $this->permission('Garden.SignIn.Allow');
+
+        $this->idParamSchema('in');
+        $in = $this->discussionPutCanonicalSchema('in')->setDescription('Set canonical url for a discussion.');
+        $out = $this->schema($this->discussionSchema(), 'out');
+
+        $body = $in->validate($body);
+
+        $row = $this->discussionByID($id);
+        if (!empty($row['Attributes']['CanonicalUrl'] ?? '')) {
+            throw new \Garden\Web\Exception\ClientException('Canonical url already set for this discussion.', 409);
+        };
+        $categoryID = $row['CategoryID'];
+        if ($row['InsertUserID'] !== $this->getSession()->UserID) {
+            $this->discussionModel->categoryPermission('Vanilla.Discussions.Edit', $categoryID);
+        }
+
+        $attributes = $row['Attributes'];
+        $attributes['CanonicalUrl'] = $body['canonicalUrl'];
+        $this->discussionModel->setProperty($id, 'Attributes', dbencode($attributes));
```

Comment:
Pretty sure this will blow away anything existing in that attributes column? @initvector do we have a standard way of handling this?